erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py

```diff
@@ -627,13 +635,47 @@ def make_item_gl_entries(self, gl_entries):
 								expense_account = service_received_but_not_billed_account
 
 					if not self.is_internal_transfer():
-						gl_entries.append(self.get_gl_dict({
-								"account": expense_account,
-								"against": self.supplier,
-								"debit": amount,
-								"cost_center": item.cost_center,
-								"project": item.project or self.project
-							}, account_currency, item=item))
+						# check if the exchange rate has changed
+						purchase_receipt_name = item.purchase_receipt
+						purchase_receipt = frappe.get_doc('Purchase Receipt', purchase_receipt_name)
+						if self.conversion_rate == purchase_receipt.conversion_rate:
+							gl_entries.append(self.get_gl_dict({
+									"account": expense_account,
+									"against": self.supplier,
+									"debit": amount,
+									"cost_center": item.cost_center,
+									"project": item.project or self.project
+								}, account_currency, item=item))
+		
+						else:
+							gle = frappe.get_all('GL Entry')
+							for gl_entry_name in gle:
+								gl_entry = frappe.get_doc('GL Entry', gl_entry_name)
+								if gl_entry.voucher_no == purchase_receipt.name:
+									if gl_entry.account == item.expense_account:
+										debit_at_old_exchange_rate = gl_entry.credit
+										break
+							gl_entries.append(
+								self.get_gl_dict({
+									"account": expense_account,
+									"against": self.supplier,
+									"debit": debit_at_old_exchange_rate,
+									"cost_center": item.cost_center,
+									"project": item.project or self.project
+								}, account_currency, item=item)
+							)
+							discrepancy_caused_by_exchange_rate_difference = amount - debit_at_old_exchange_rate
```

Comment:
No need to query GL Entries for difference amount. This you can calculate using (item.qty * item.rate) * (purchase_receipt_converison_factor - purchase_invoice_conversion_factor)

Also no need for if else block. Only if condition should be sufficient. If purchase_receipt_converison_factor != purchase_invoice_conversion_factor then just add 2 additional entries for difference amount

Reply:
Done